### DISCUSSION TITLE
General Understanding

### DISCUSSION THREAD

**Discussion by AndiLavera** [react-static/react-static#1521] [Q&A]
Hello react-static team, first let me say thank you for this awesome library!

I am moving from nextjs to react-static. I have a few questions related to my lack of understanding. My use case is more of a CRUD app than anything else. I have it connected with a rails api server. 

### Dynamic Routes
What actually is a dynamic route? Let's say we have a pet CRUD app. Index shows all the pets and would probably be `/pets`. The show view is probably `/pets/:id`. Ignoring the fact the page is 98% the same (excluding input values or something), this is a dynamic route?

I actually am testing with this pet idea. Currently, I am not excluding `/pets/:id` from the prefetch excludes. Should I be?

### Initial props
Nextjs has 2 nice hooks called `getInitialProps` and `getServerSideProps`. Does react-static have anything of the sorts? On my pets show page, I am just using useEffect and fetching the proper pet based on the supplied `id`. Is this the best choice or is there something else? More specifically, is there something that would get ran server side so the client receives the HTML with the data? I don't believe this is a good candidate for routeData as there may be thousands of pets or other objects? Is this correct? Could I possibly throw an axios get request into the `getData` for that page? 

I just bundled for production and ran serve, it seems like my toy app is working. The pets show page I have been talking about did get bundled, then on the frontend, the useEffect is kicking in and supplying the proper data? Does this mean I am using react-static right or is there something I am missing that would make it more performant?

**Comment by SleeplessByte** [comment]
A dynamic route is a route that you can't determine at build time. So if you have a CRUD app, and you are going to static-ly generate all the pages, you'll know all the pets, and thus their ids, at build time, and so it's not dynamic. You only need to exclude paths that are not generated by react-static (because they won't have a generated `routeInfo.json`).

You generally would push the "intial data" (build-time data) to `routeData`. It then becomes available via `useRouteData()`. You _may_ choose to fetch the "fresh" data from a server and then replace something. So each page would only have the required data for _that_ page.

**Reply by AndiLavera** [reply]
That all makes sense. Thank you for the nice explanation. 

The only question I have left is if react-static supports something like useEffect/getInitialProps server-side. Keeping with my example, we will not know all the pets at build-time. I mean we could but it could be thousands of pets with new pets being added every second. I figured the next best thing was to fire off the useEffect/getInitialProps server-side and then hand over the completed html like you can do in nextjs.

**Reply by Pixelatex** [reply]
if you have a website that is highly dynamic with new content added every second... you might want to reconsider if static is what you want to use. An SSR solution like next is actually more suited to this use-case. Correct me if wrong @SleeplessByte but wasn't the dynamic routes in RS only an additional feature and not the core business? It's a bit like using your car as a clothes closet, sure it would technically work but it's far from the most practical solution

**Reply by AndiLavera** [reply]
@Pixelatex Oh got it! That would make sense. I figured where nextjs worked, react-static would also work.

**Reply by SleeplessByte** [reply]
Absolutely @Pixelatex . 

@andrewc910 We _recommend_ nextjs for something like that!
